File: src/Commands/Model/Graph/GraphException.cs
using System;
using System.Collections.Generic;

namespace PnP.PowerShell.Commands.Model.Graph
{
    public class GraphException : Exception
    {
        public GraphError Error { get; set; }

        public string AccessToken { get; set; }

        public System.Net.Http.HttpResponseMessage HttpResponse { get; set; }
    }

    public class GraphError
    {
        public string Code { get; set; }

        public string Message { get; set; }

        public GraphError InnerError { get; set; }

        public Dictionary<string, object> AdditionalData { get; set; }

        public string ThrowSite { get; set; }
    }
}

File: pnp-cli/src/models/graph/graph-exception.ts
export interface GraphError {
  code?: string;
  message?: string;
  innerError?: {
    'request-id'?: string;
    date?: string;
    code?: string;
  };
}

export class GraphException extends Error {
  error?: GraphError;
  accessToken?: string;
  httpStatusCode?: number;
  httpResponse?: Response;

  constructor(error?: GraphError, statusCode?: number) {
    super(error?.message || 'Microsoft Graph API error');
    this.name = 'GraphException';
    this.error = error;
    this.httpStatusCode = statusCode;
  }

  static fromResponse(body: string, statusCode: number, response?: Response): GraphException {
    try {
      const parsed = JSON.parse(body);
      const error = parsed.error as GraphError | undefined;
      const exception = new GraphException(error, statusCode);
      exception.httpResponse = response;
      return exception;
    } catch {
      const exception = new GraphException({ message: body }, statusCode);
      exception.httpResponse = response;
      return exception;
    }
  }
}
